pennylane/dla/lie_closure.py

```diff
@@ -0,0 +1,274 @@
+        Checks if ``pauli_sentence`` is linearly independent.
+
+        This is done in the following way: ``M`` (see description in class) is extended by ``pauli_sentence``.
+        If the added operator has a PauliWord (key) that is new to ``pw_to_idx``, then we have to add a new row
+        and already know that it has to be linearly independent.
+        If it contains the same PauliWords, we need to compute the new rank and compare it with the old rank.
+        If the rank is the same, the operator is linearly dependent and not added. Else, the rank is incrased by 1
+        and the extended M becomes our new M.
+
+        Args:
+            M (ndarray): coefficient matrix for current LIS
+            pauli_sentence (`~.PauliSentence`): Pauli sentence for which to add a column if independent
+            pw_to_idx (dict): map from ``PauliWord`` to row index in ``M``
+            rank (int): current rank of ``M``, equal to its number of columns
+            num_pw (int): current number of ``PauliWord``\ s, equal to the number of rows in ``M``
+        Returns:
+            ndarray: updated coefficient matrix for the LIS
+            dict: updated map from ``PauliWord`` to row index in ``M``. Includes new ``PauliWord`` keys
+                from the input ``pauli_sentence`` if it was linearly independent
+            int: updated rank/number of columns of ``M``
+            int: updated number of ``PauliWord``\ s/number of rows of ``M``
+            bool: whether ``pauli_sentence`` was linearly independent and its column was added to ``M``
+        """
+        new_pws = [pw for pw in pauli_sentence.keys() if pw not in pw_to_idx]
+        new_num_pw = num_pw + len(new_pws)
+
+        if new_num_pw < rank + 1:
+            # Can't span rank+1 independent vectors in fewer than rank+1 dimensions
+            # The input PauliSentence must have been linearly dependent
+            return M, pw_to_idx, rank, num_pw, False
+
+        M = np.pad(M, ((0, new_num_pw - num_pw), (0, 1)))
+
+        # If there are new PauliWords (i.e. new basis vectors), the candidate vector must be linearly independent
+        if new_num_pw > num_pw:
+            new_pw_to_idx = copy(pw_to_idx)
+            for i, pw in enumerate(new_pws, start=num_pw):
+                new_pw_to_idx[pw] = i
+            # Add new PauliSentence entries to matrix
+            for pw, value in pauli_sentence.items():
+                M[new_pw_to_idx[pw], rank] = value
+
+            return M, new_pw_to_idx, rank + 1, new_num_pw, True
+
+        # Add new PauliSentence entries to matrix
+        for pw, value in pauli_sentence.items():
+            M[pw_to_idx[pw], rank] = value
+
+        # Check if new vector is proportional to any of the previous vectors
+        # This is significantly cheaper than computing the rank and should be done first
+        if _is_any_col_propto_last(M):
+            M = M[:num_pw, :rank]
+            return M, pw_to_idx, rank, num_pw, False
```

Comment:
I didnt know `np.linalg.cholesky`, looks neat!
While playing around with it I get a lot of cases where it complains that the its input (`np.conj(M.T) @ M`) is not positive definite, e.g.

```python3
M = np.array(
    [
        [1.0, 0.0, 0.5, 0.5, 1.0],  # matching 2nd to last and last
        [1.0, 0.5, 0.0, 0.0, 0.0],
        [1.0, 0.5, 1.0, 2.0, 4.0],
    ]
)
np.linalg.cholesky(np.conj(M.T) @ M)
```

```pycon
---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
..
----> 8 np.linalg.cholesky(np.conj(M.T) @ M)

File ~/anaconda3/envs/pennylane311/lib/python3.11/site-packages/numpy/linalg/linalg.py:779, in cholesky(a)
    777 t, result_t = _commonType(a)
    778 signature = 'D->D' if isComplexType(t) else 'd->d'
--> 779 r = gufunc(a, signature=signature, extobj=extobj)
    780 return wrap(r.astype(result_t, copy=False))

File ~/anaconda3/envs/pennylane311/lib/python3.11/site-packages/numpy/linalg/linalg.py:115, in _raise_linalgerror_nonposdef(err, flag)
    114 def _raise_linalgerror_nonposdef(err, flag):
--> 115     raise LinAlgError("Matrix is not positive definite")

LinAlgError: Matrix is not positive definite

Reply:
Yes, if `np.linalg.cholesky(np.conj(M.T) @ M)` is not positive definite it means the columns are linearly dependent. That's why it should be put in a `try..except` block. If it raises, throw away the new column, otherwise include it.

Reply:
Not sure whether this will be relevant (but maybe yes, because we call the function repeatedly), but `try...except` is usually quite slow if exceptions are expected part of the execution flow, iirc.

Reply:
More general question: what is the advantage of keeping M orthonormalized aside from cleaner code? Would you reckon it becomes faster or slower overall (in particular for large VSpaces)?

Reply:
> I would say we can keep M non-orthonormal and if we see meaningful opportunities for optimization in practice

I think that is a good idea. I see the main advantage in possibly changing the basis in adding multiple operators at once (which is necessary in rounds in `lie_closure` where many candidates are computed (see https://github.com/PennyLaneAI/pennylane/pull/5169) - so instead of going through them one by one the maximal linearly-independent subset could be chosen among all of them + former ops at once.